tracer/test/Datadog.Trace.Tests/Agent/MessagePack/SpanMessagePackFormatterTests.cs

```diff
@@ -208,6 +208,156 @@ public void SpanLink_Tag_Serialization()
+        var mockApi = new MockApi();
+        var settings = TracerSettings.Create(new());
+        var agentWriter = new AgentWriter(mockApi, statsAggregator: null, statsd: null, automaticFlush: false);
+        var tracer = new Tracer(settings, agentWriter, sampler: null, scopeManager: null, statsd: null, NullTelemetryController.Instance, discoveryService);
+
+        tracer.TracerManager.Start();
+        discoveryService.TriggerChange(spanEvents: nativeSpanEventsEnabled);
+        var formatter = SpanFormatterResolver.Instance.GetFormatter<TraceChunkModel>();
+
+        var parentContext = new SpanContext(new TraceId(0, 1), 2, (int)SamplingPriority.UserKeep, "ServiceName1", "origin1");
+        var traceContext = new TraceContext(tracer);
+        var spanContext = new SpanContext(parentContext, traceContext, "ServiceName1");
+        var span = new Span(spanContext, DateTimeOffset.UtcNow);
+
+        var eventName = "test_event";
+        var eventTimestamp = new DateTimeOffset(2024, 1, 1, 0, 0, 0, TimeSpan.Zero);
+        var eventAttributes = new List<KeyValuePair<string, object>>
+        {
+            new("string_key", "hello"),
+            new("empty_value", null),
+            new("bool_key", true),
+            new("int_key", 42),
+            new("double_key", 3.14),
+            new("string_array", new[] { "item1", "item2", "item3" }),
+        };
+
+        var eventAttributes2 = new List<KeyValuePair<string, object>>
+        {
+            new("object_array", new object[] { "string", 42, true }),
+            new("bool_array", new[] { true, false, true }),
+            new("int_array", new[] { 123, 1234, 12345 }),
+            new("double_array", new[] { 1.2, 1.3210, 200000.1 }),
+            new(null, "empty_key"),
+            new("string_key", "hello"),
+        };
+
+        span.AddEvent(new SpanEvent(eventName, eventTimestamp, eventAttributes));
+        span.AddEvent(new SpanEvent("another_event", eventTimestamp.AddSeconds(1), eventAttributes2));
+        span.SetDuration(TimeSpan.FromSeconds(1));
+
+        var traceChunk = new TraceChunkModel(new([span]));
+        byte[] bytes = [];
+        var length = formatter.Serialize(ref bytes, 0, traceChunk, SpanFormatterResolver.Instance);
+        var result = global::MessagePack.MessagePackSerializer.Deserialize<MockSpan[]>(new ArraySegment<byte>(bytes, 0, length));
+
+        result.Should().HaveCount(1);
+        var deserializedSpan = result[0];
+
+        if (nativeSpanEventsEnabled)
+        {
+            deserializedSpan.SpanEvents.Should().HaveCount(2);
+
+            var firstEvent = deserializedSpan.SpanEvents[0];
+            firstEvent.Name.Should().Be(eventName);
+            firstEvent.Timestamp.Should().Be(eventTimestamp.ToUnixTimeNanoseconds());
+
+            var attributes = firstEvent.Attributes;
+
+            attributes["string_key"].Type.Should().Be(0);
+            attributes["string_key"].StringValue.Should().Be("hello");
```

Comment:
nit: I think it would be valuable to make additional assertions such as `attributes["string_key"].BoolValue.Should().BeNull()` to make sure there's only one value being set on each attribute